### PR TITLE
Add Nhibernate as a source

### PR DESCRIPTION
**Pull request by WojcikMike**

### CODE REVIEW

**Review by WilliamBZA — APPROVED** [review]

File: src/TimeoutMigrationTool.NHibernate.AcceptanceTests-Oracle/NHibernateAcceptanceTests.cs [hunk 1]
```diff
@@ -0,0 +1,74 @@
...
+    using System;
+    using System.Collections.Generic;
+    using System.Text;
+    using System.Threading.Tasks;
+
+    class NHibernateAcceptanceTests
+    {
+        public NHibernateAcceptanceTests()
+        {
+            connectionString = $@"Data Source = (DESCRIPTION = (ADDRESS_LIST = (ADDRESS = (PROTOCOL = TCP)(HOST = 127.0.0.1)(PORT = 1521)))(CONNECT_DATA = (SERVER = DEDICATED)(SERVICE_NAME = ORCLCDB.localdomain))); DBA Privilege = SYSDBA; User Id = sys; Password = Oradoc_db1; Enlist = dynamic";
```
Comment (WilliamBZA) [inline src/TimeoutMigrationTool.NHibernate.AcceptanceTests-Oracle/NHibernateAcceptanceTests.cs:19]: Assumes the default username/password for oracle docker images

File: src/TimeoutMigrationTool/NHibernate/NHibernateTimeoutsSource.cs [hunk 2]
```diff
@@ -0,0 +1,300 @@
...
+
+                var stagedTimeoutEntities = await session.QueryOver<StagedTimeoutEntity>()
+                    .Where(stagedTimeoutEntity => stagedTimeoutEntity.BatchState != BatchState.Completed)
+                    .SelectList(timeoutEntities => timeoutEntities.SelectGroup(te => te.BatchNumber)
+                        .SelectGroup(te => te.BatchState)
+                        .SelectCount(te => te.Id))
+                    .OrderBy(entity => entity.BatchNumber)
+                    .Asc.ListAsync<object[]>();
+
+                return stagedTimeoutEntities.Select(x => new BatchInfo((int) x[0], (BatchState) x[1], (int) x[2]))
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/NHibernate/NHibernateTimeoutsSource.cs:70]: I was wondering if we could use Nhibernates way of projecting into non-tracked entities here and if that would make it more readable. Not against this approach by the way and maybe it is not worth the extra effort. Just a thought

File: src/TimeoutMigrationTool/NHibernate/NHibernateTimeoutsSource.cs [hunk 3]
```diff
@@ -0,0 +1,300 @@
...
+
+            return timeouts.Select(x => new EndpointInfo
+                {
+                    EndpointName = x[0].ToString(),
+                    LongestTimeout = (DateTime)x[1],
+                    ShortestTimeout = (DateTime)x[2],
+                    NrOfTimeouts = (int)x[3],
+                    Destinations = session.QueryOver<TimeoutEntity>()
+                                            .Where(timeout => timeout.Endpoint == x[0].ToString())
+                                            .SelectList(t => t.SelectGroup(x => x.Destination))
```
Comment (danielmarbach) [inline src/TimeoutMigrationTool/NHibernate/NHibernateTimeoutsSource.cs:267]: I was wondering if we could use Nhibernates way of projecting into non-tracked entities here and if that would make it more readable. Not against this approach by the way and maybe it is not worth the extra effort. Just a thought

**Review by danielmarbach — APPROVED** [review]

File: src/TimeoutMigrationTool/NHibernate/NHibernateDialectValidator.cs [hunk 4]
```diff
@@ -0,0 +1,28 @@
+﻿namespace Particular.TimeoutMigrationTool.NHibernate
+{
+    using System;
+    using McMaster.Extensions.CommandLineUtils;
+    using McMaster.Extensions.CommandLineUtils.Validation;
+    using System.ComponentModel.DataAnnotations;
+
+    public class NHibernateDialectValidator : IOptionValidator
```
Comment (lailabougria) [inline src/TimeoutMigrationTool/NHibernate/NHibernateDialectValidator.cs:8]: This is nice, we should check if we can do the same for the ravenversion parameter. That is, if we keep raven 3 around.

**Review by lailabougria — APPROVED** [review]

**Pull request merged**